src/ensemble/src/main/scala/EnsembleByKey.scala

```diff
@@ -0,0 +1,194 @@
+  def getColNames: Array[String] = $(colNames)
+
+  def setColNames(arr: Array[String]): this.type = set(colNames, arr)
+
+  def setColNames(arr: String*): this.type = set(colNames, arr.toArray)
+
+  def setColName(value: String): this.type = set(colNames, Array(value))
+
+  val allowedStrategies = Set("mean")
+  val strategy = StringParam(this, "strategy", "How to ensemble the scores ex:" +
+    " mean", { x: String => allowedStrategies(x) })
+
+  def getStrategy: String = $(strategy)
+
+  def setStrategy(value: String): this.type = set(strategy, value)
+
+  setDefault(strategy -> "mean")
+
+  val collapseGroup = BooleanParam(this, "collapseGroup", "whether to collapse all items in group to one entry")
+
+  def getCollapseGroup: Boolean = $(collapseGroup)
+
+  def setCollapseGroup(value: Boolean): this.type = set(collapseGroup, value)
+
+  setDefault(collapseGroup -> true)
+
+  override def transform(dataset: Dataset[_]): DataFrame = {
+
+    if (get(colNames).isEmpty) {
+      setDefault(colNames -> getCols.map(name => s"$getStrategy($name)"))
+    }
+
+    transformSchema(dataset.schema)
+
+    val strategyToFloatFunction = Map(
+      "mean" -> { (x: String, y: String) => mean(x).alias(y) }
+    )
+
+    val strategyToVectorFunction = Map(
+      "mean" -> { (x: String, y: String) =>
+        val dim = dataset.select(x).take(1)(0).getAs[DenseVector](0).size
+        new VectorAvg(dim)(dataset(x)).asInstanceOf[Column].alias(y)
+      }
+    )
+
+    val newCols = getCols.zip(getColNames).map { case (inColName, outColName) =>
+      dataset.schema(inColName).dataType match {
+        case _: DoubleType =>
+          strategyToFloatFunction(getStrategy)(inColName, outColName)
+        case _: FloatType =>
+          strategyToFloatFunction(getStrategy)(inColName, outColName)
+        case v if v == VectorType =>
+          strategyToVectorFunction(getStrategy)(inColName, outColName)
+        case t =>
+          throw new IllegalArgumentException(s"Cannot operate on type:$t with strategy:$getStrategy")
+      }
+    }
+
+    val aggregated = dataset.toDF()
+      .groupBy(getKeys.head, getKeys.tail: _*)
+      .agg(newCols.head, newCols.tail: _*)
+
+    if (getCollapseGroup) {
+      aggregated
+    } else {
+      val needToDrop = getColNames.toSet & dataset.columns.toSet
+      dataset.drop(needToDrop.toList: _*).toDF().join(aggregated, getKeys)
+    }
+
+  }
+
+  def transformSchema(schema: StructType): StructType = {
+    val colSet = getCols.toSet
+    val colToNewName = getCols.zip(getColNames).toMap
+
+    val newFields = schema.fields.flatMap { f =>
+      if (colSet(f.name)) {
+        val newField = StructField(colToNewName(f.name), f.dataType)
+        f.dataType match {
+          case _: DoubleType => Some(newField)
+          case _: FloatType => Some(newField)
+          case fdt if fdt == VectorType => Some(newField)
+          case t => throw new IllegalArgumentException(s"Cannot operate on type:$t with strategy:$getStrategy")
```

Comment:
This already happens on line 77. I need to repeat some code, because actually applying the funtions requires knowing the dimensionality of the vector column which is only available at transform time